File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution(object):
    def __init__(self):
                self.visited = {}
        
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        
        
        if node is None:
            return None
        
        if node in self.visited:
            return self.visited[node]
        
        new_node = Node(node.val, [])
        
        self.visited[node] = new_node
        
        if node.neighbors:
            new_node.neighbors = [self.cloneGraph(n) for n in node.neighbors]
        
        return new_node
        
    

File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        HashMap<Node, Node> oldToNew = new HashMap<>();
        return node != null ? clone(node, oldToNew) : null;
    }

    private Node clone(Node node, HashMap<Node, Node> oldToNew) {
        if (oldToNew.containsKey(node)) {
            return oldToNew.get(node);
        }

        Node newNode = new Node(node.val);
        oldToNew.put(node, newNode);
        for (Node neighbor: node.neighbors) {
            newNode.neighbors.add(clone(neighbor, oldToNew));
        }

        return newNode;
    }
}